### PR TITLE
Get Previous Block Root From State

### LINKED ISSUE #1980: Use Previous Block Root From State instead of Saved Block
Currently we retrieve the chain head and hash it to generate the previous block root for state transitions. We should instead be using the previous block root from the state. This would solve some edge cases for us in sync. Ref: https://github.com/ethereum/eth2.0-specs/issues/692#issuecomment-472346984

### PR DESCRIPTION
**Pull request by nisdas**
This resolves #1980, where we retrieve block root from the state instead of performing a read for it from the db. This also resolves issues with initial sync in the case of skipped slots.

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -8,6 +8,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+
```
Comment (terenc3t) [inline beacon-chain/blockchain/service.go]: we should remove the spacing here? stupid IDE... 😞

Comment (prestonvanloon) [reply beacon-chain/blockchain/service.go]: yes

**Comment by nisdas** [comment]
Don't Merge this in yet, its dependent on {{2006}} getting merged in first

**Comment by terenc3t** [comment]
> Don't Merge this in yet, its dependent on #2006 getting merged in first

Why does this depend on #2006?

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -186,17 +187,12 @@ func (c *ChainService) StateInitializedFeed() *event.Feed {
 
 // ChainHeadRoot returns the hash root of the last beacon block processed by the
 // block chain service.
-func (c *ChainService) ChainHeadRoot() ([32]byte, error) {
-	head, err := c.beaconDB.ChainHead()
+func (c *ChainService) ChainHeadRoot(state *pb.BeaconState) ([32]byte, error) {
+	root, err := b.BlockRoot(state, state.Slot-1)
```
Comment (terenc3t) [inline beacon-chain/blockchain/service.go]: we should be careful of this, imagine the following the scenario:
**validator current time:** 25s
**beacon node state slot:** 3
Beacon node has just processed the head for slot 3, validator wants whatever the head is at slot 3 but given `state.Slot-1`, validator can only get the block root at slot 2 but it should be 3

Comment (rauljordan) [reply beacon-chain/blockchain/service.go]: Hey @nisdas is this still relevant?

Comment (rauljordan) [reply beacon-chain/blockchain/service.go]: Bump @nisdas

**Comment by nisdas** [comment]
Not relevant anymore

**Pull request closed without merging**